drivers/vulkan/rendering_device_vulkan.cpp

```diff
@@ -7684,6 +7687,9 @@ void RenderingDeviceVulkan::_draw_list_free(Rect2i *r_last_viewport) {
 		memdelete(draw_list);
 		draw_list = nullptr;
 	}
+
+	// draw_list is no longer active
+	_THREAD_SAFE_UNLOCK_
```

Comment:
unlock is not needed

Reply:
`_THREAD_SAFE_LOCK_` isn't a MutexLock, it's a `.lock()` call, so we need to be sure to `unlock()` it when the draw list is complete.

Reply:
I think there is a mix up here, 99% of the code base uses `_THREAD_SAFE_METHOD_` which makes sense to use here as well. `_THREAD_SAFE_METHOD_` automatically unlocks after the method is called (it uses a temporary variable that when it goes out of scope unlocks). It's a safer approach then using `_THREAD_SAFE_LOCK_` as you guarantee the lock being freed, while `_THREAD_SAFE_LOCK_` introduces the risk of exiting the method early and not unlocking.

Reply:
Ah hang on, sorry guys, I misread this. It's two methods, `_draw_list_allocate`, in which we lock, and `_draw_list_free` in which we unlock. Then it does make sense to use `_THREAD_SAFE_LOCK_` and `_THREAD_SAFE_UNLOCK_` though. 

That said, it feels dangerous to me. There is a lot of potential here for a mismatch between lock and unlock. That one I'll leave to @reduz to comment on.

Reply:
The problem is that thread safety must be maintained for the entire duration that the draw list is being built, not just in the context of these individual functions.

I understand the risk of manual locking/unlocking. We could maybe modify the draw_list to hold a MutexLock for the duration of its life, but that would have equivalent risk.

Reply:
If at any point a draw list is created and not ended, you likely have other serious problems in the renderer anyways.

Reply:
Yup, agree @brianwinterpixel , I think @reduz just mixed up the two approaches.

I haven't examined the code base close enough for 2D rendering, mostly been working on 3D and we call other calls to create draw lists, but they internally may end up calling the method you changed. I guess its that sort of scenarios I would caution about when lock and unlock are loosely joined.

I don't really have a more valid alternative though, I think @reduz is best to comment here